crates/forge-runner/src/gas.rs

```diff
@@ -1,79 +1,164 @@
+    let messaging_gas_vector = get_messages_costs(
+        &resources.l2_to_l1_payloads_lengths,
+        &resources.l1_handler_payloads_lengths,
+    );
+
+    let l1_and_vm_costs = get_l1_and_vm_costs(
+        l1_data_cost,
+        transaction_context,
+        resources.execution_resources,
+    );
+
+    let gas = l1_and_vm_costs + messaging_gas_vector;
+
+    Ok(gas.l1_gas)
+}
+
+// Put together from a few blockifier functions
+fn get_messages_costs(
+    l2_to_l1_payloads_lengths: &[usize],
+    l1_handler_payloads_lengths: &[usize],
+) -> GasVector {
+    let l2_to_l1_segment_length = l2_to_l1_payloads_lengths
+        .iter()
+        .map(|payload_length| constants::L2_TO_L1_MSG_HEADER_SIZE + payload_length)
+        .sum::<usize>();
+
+    let l1_to_l2_segment_length = l1_handler_payloads_lengths
+        .iter()
+        .map(|payload_length| constants::L1_TO_L2_MSG_HEADER_SIZE + payload_length)
+        .sum::<usize>();
+    let message_segment_length = l2_to_l1_segment_length + l1_to_l2_segment_length;
+
+    let n_l2_to_l1_messages = l2_to_l1_payloads_lengths.len();
+    let n_l1_to_l2_messages = l1_handler_payloads_lengths.len();
+
+    let l1_handlers_emission_costs = l1_handler_payloads_lengths
+        .iter()
+        .map(|l1_handler_payload_length| {
+            get_consumed_message_to_l2_emissions_cost(Some(*l1_handler_payload_length))
+        })
+        .sum();
+    let starknet_gas_usage = GasVector {
+        l1_gas: u128_from_usize(
+            message_segment_length * eth_gas_constants::GAS_PER_MEMORY_WORD
+                + n_l2_to_l1_messages * eth_gas_constants::GAS_PER_ZERO_TO_NONZERO_STORAGE_SET
+                + n_l1_to_l2_messages * eth_gas_constants::GAS_PER_COUNTER_DECREASE,
+        )
+        .expect("Could not convert starknet gas usage from usize to u128."),
+        l1_data_gas: 0,
+    } + get_log_message_to_l1_emissions_cost(l2_to_l1_payloads_lengths)
+        + l1_handlers_emission_costs;
+
+    let sharp_gas_usage = GasVector {
+        l1_gas: u128_from_usize(
+            message_segment_length * eth_gas_constants::SHARP_GAS_PER_MEMORY_WORD,
+        )
+        .expect("Could not convert sharp gas usage from usize to u128."),
+        l1_data_gas: 0,
+    };
+
+    starknet_gas_usage + sharp_gas_usage
+}
+
+fn get_l1_and_vm_costs(
+    l1_data_costs: GasVector,
+    transaction_context: &TransactionContext,
+    execution_resources: ExecutionResources,
+) -> GasVector {
+    let versioned_constants = transaction_context.block_context.versioned_constants();
+    let resources_mapping = get_resources_mapping(l1_data_costs, execution_resources);
+
+    calculate_tx_gas_vector(&resources_mapping, versioned_constants)
+        .expect("Could not calculate gas")
+}
+
+fn add_syscall_resources(transaction_context: &TransactionContext, resources: &mut UsedResources) {
+    let versioned_constants = transaction_context.block_context.versioned_constants();
```

Comment:
[nit] `versioned_constans` could be passed to `add_syscall_resources` and `get_l1_and_vm_costs` instead of `transaction_context` so we don't call the function twice